### ISSUE TITLE
Shuffle always starts with first track

### ISSUE THREAD

**Issue by Ghostbird** [bcc-code/bmm-web#218]
https://github.com/bcc-code/bmm-web/blob/284ec84af2ad5d40fcd3059ac16a0a5510f49a77/plugins/mediaPlayer/Queue.ts#L47

The shuffle function leaves the currently playing track in position. Which makes some sense at a certain level. However, the definition of current track defaults to the first track as soon as a `Queue` is initialised. Hence when you click shuffle, it shuffles only the rest of the list.

I think it would be better for now, to not exempt the current track from the shuffle. Fancy features that handle edge cases can come later, when we _really_ know exactly when and how to handle them.

**Comment by SimonSimCity** [comment]
The current implementation is in line with most of the big players in this market, at least with Youtube Music and Spotify, why I'd like to know a more solid reason for changing this, if this actually is what you are asking for.

As far as I understand your issue, you want to shuffle the list before you start playing. This is a use-case which simply has not been implemented yet. I don't see how this would be in conflict with anything existing, as it sounds as of your description, but rather an additional feature, which needs to be implemented.

**Comment by SimonSimCity** [comment]
If you are referring to the `Shuffle` button on the page for curated playlists, this button was not implemented correctly by me, that's right, but the place to be changed is not in the function you're pointing out, but is because of a missing feature, which should be used instead.

**Comment by Ghostbird** [comment]
Apologies, is there another shuffle button? I guess I looked at the wrong function then. Strange that it would explain exactly this bug.

That aside, I think shuffling a list before you start playing is the common use case, but I have only a single data point to back this up. I wouldn't even know how to define "shuffling a currently playing list". I normally shuffle a list before playing it, or stop playing it, shuffle it, then play it.

My suggestion for changing it was simply this:

- I observe that the current behaviour is incorrect.
- I observe that there is some exceptional behaviour beyond a simple shuffle algorithm is programmed that causes this behaviour.
- I suggest to remove this exception, which would remove the incorrect behaviour, until such a time where that exception can be implemented in a way that does not cause the incorrect behaviour.

Personally I prefer limited functionality over perceived wrong functionality. However, if this code is called by another button as you state, than that perception may be wrong.

**Comment by Ghostbird** [comment]
I've found the other shuffle button, the UI is still not finished, so I hadn't seen it, as it ended up in a weird place.
I discussed this with a colleague and we looked at Spotify, and I propose that if you want to exclude the currently playing track in the shuffle function it should do this:

- Take the currently playing track out of the list
- Shuffle the list
- Prepend the currently playing track

This seems to be what Spotify does, and it matches my understanding of a shuffle. When you shuffle, you so before playing the list. Hence when you want to continue playing the current track, you actually want to a list that consists of all tracks _except_ the currently playing track. Then the current track must be the new starting point of the combined queue.

The current shuffle function does it in a very odd way:

Imagine I have a playlist:

[ A, B, C, D ]
I've listened for a while and am at track C.

In the current shuffle it retains the _index_ of C.
I'd end up playing e.g. [ D, B, C ,A ] and would at the end of the list never have heard track D.

In Spotify the _identity_ of the currently playing track is retained, but its index is reset to 0
[ C, B, D, A ]
If you finished listening that, you'd not have "missed" any tracks in the list.
